### PR TITLE
Add user content translations with configurable backends

### PR DESCRIPTION
**Pull request by Gargron**
Server owners can configure a translation backend that enables users to translate public posts. Results are cached, so requests for translating the same string into the same language, even across different posts, does not hit the backend more than once.

### Supported backends

- DeepL (`DEEPL_API_KEY` and `DEEPL_PLAN`)
- LibreTranslate (`LIBRE_TRANSLATE_ENDPOINT` and `LIBRE_TRANSLATE_API_KEY`)

### Screenshots

|Original|Translated|
|--|--|
|image|image|

### REST API

- `POST /api/v1/statuses/:id/translate` (supports global `lang` param, otherwise defaults to user's locale), returns `content` and `detected_source_language`. May return HTTP 404 if no backend is configured, or HTTP 503 if the backend fails to translate.

___

This project was funded through the [NGI0 Discovery](https://nlnet.nl/discovery) Fund, a fund established by [NLnet](https://nlnet.nl/) with financial support from the European Commission's [Next Generation Internet](https://ngi.eu/) programme, under the aegis of DG Communications Networks, Content and Technology under grant agreement No 825322.

### CODE REVIEW

File: app/lib/translation_service/deepl.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      else
+        raise UnexpectedResponseError
+      end
+    end
+  end
+
+  private
+
+  def request(text, source_language, target_language)
+    req = Request.new(:post, 'https://api-free.deepl.com/v2/translate', form: { text: text, source_lang: source_language.upcase, target_lang: target_language, tag_handling: 'html' })
```
Comment (ykzts) [inline app/lib/translation_service/deepl.rb:30]: Can we use an environment variable to specify the URL for DeepL API Pro?

**Review by ykzts — CHANGES_REQUESTED** [review]

File: app/lib/translation_service/deepl.rb (hunk 1 above)
Comment (ykzts) [reply app/lib/translation_service/deepl.rb:30]: FYI `api.deepl.com`

Comment (Gargron) [reply app/lib/translation_service/deepl.rb]: Done!

**Review by ykzts — APPROVED** [review]

**Comment by shleeable** [comment]
This is excellent. Can we see support for https://github.com/browsermt/mts?

**Comment by FloatingGhost** [comment]
some gotchas from akkoma's implementation that might be helpful here

not all target languages are locale codes - whilst `en` does translate for backwards compatibility, deepL recommends using `EN-GB` and `EN-US`, whilst they don't say explicitly that they'll remove `en`, it is possible since it's no longer in their "supported languages" list. 

so, you might want to provide users with a setting to select preferred target language, this covers the above, and also since deepl only works on a relatively small subset of languages - a much smaller subset than those supported by mastodon's locale set, and these users will never be able to translate which isn't great

**Comment by Tak** [comment]
What about advertising this capability to clients in `api/v1/instance` or elsewhere?

**Comment by gllmhyt** [comment]
I would like to disable this feature on the instance I'm rocking, but I cannot seem to find where to hide the button asking for translation. Is it available or is it not yet implemented? When no API key for a translation service is given, the translate button should not be displayed (as it will fail/error every time).

**Comment by Tak** [comment]
This is also my motivation for wanting capability detection in the api

**Comment by SGombert** [comment]
There is an open-source software for self-hosting translation services called EasyNMT which supports many languages and works okay (at least the opus and m2m models). https://github.com/UKPLab/EasyNMT Would it be possible also to support this backend? Unfortunately, I don't know Ruby so far, but I guess it should not be too complicated to derive code for supporting another backend from the existing one, so I will maybe contribute to this as soon as I find the time.

**Comment by vollans** [comment]
I've enabled the LibreTranslate version on my instance. A lot of foreign language posts either don't have a language set or are set as English, and so Translate isn't available. Is it possible to add a "Force Translate" to the "..." option?

**Comment by ineffyble** [comment]
@vollans Can you [raise a feature request](https://github.com/mastodon/mastodon/issues/new?assignees=&labels=suggestion&template=2.feature_request.yml)?

**Comment by JimKillock** [comment]
Did anyone think about allowing manual translations? (That is what I assumed “user translation” was when I read the feature, and to my mind would be very helpful for people writing / posting, as they may want multilingual content, but prefer to have accurate human translations).

**Comment by MasinAD** [comment]
> There is another open-source software for self-hosting translation services called EasyNMT which supports many languages and uses more state-of-the-art techniques than LibreTranslate. https://github.com/UKPLab/EasyNMT Would it be possible also to support this backend? Unfortunately, I don't know Ruby so far, but Iooked at the code for the existing backends and it looks straightforward, so I will probably develop one for EasyNMT as soon as I find time.

May I refer you to 
grafik
(Source: [Argos Translate](https://github.com/argosopentech/argos-translate/) )

**Comment by MasinAD** [comment]
But I came here for something else: What's the best way to ask for support of [translateLocally](https://github.com/XapaJIaMnu/translateLocally)? Should I [raise a feature request](https://github.com/mastodon/mastodon/issues/new?assignees=&labels=suggestion&template=2.feature_request.yml)?

**Pull request merged**